### PR TITLE
Memory management and other small fixes

### PR DESCRIPTION
**Pull request by bennywwg**
Added release statements to most allocated objects. Memory leaks are essentially eliminated, primarily by releasing command buffers which are created each frame.

### CODE REVIEW

File: source/ngf-mtl/impl.cpp [hunk 1]
```diff
@@ -448,6 +465,64 @@ static uint32_t ngfmtl_get_num_rows(const uint32_t height, const ngf_image_forma
...
+
+  ngf_id() : ptr_(nullptr) {
+  }
+  ngf_id(const ngf_id_init_types& type) : ptr_(T::alloc()->init()) {
+  }
+  // Note: Does NOT increment ref count. You can use this directly after calling alloc()->init()
+  ngf_id(T* starting_ptr) : ptr_(starting_ptr) {
+  }
+  ~ngf_id() {
+    if (ptr_) { ptr_->release(); }
```
Comment (nicebyte) [inline source/ngf-mtl/impl.cpp:490]: i usually separate this de-initialization code into a `destroy_if_necessary` method and call it from both the dtor and copy/move assignment operators.

File: source/ngf-mtl/impl.cpp [hunk 2]
```diff
@@ -448,6 +465,64 @@ static uint32_t ngfmtl_get_num_rows(const uint32_t height, const ngf_image_forma
...
+  }
+
+  ngf_id(const ngf_id& other) : ptr_(other.ptr_) {
+    if (ptr_) { ptr_->retain(); }
```
Comment (nicebyte) [inline source/ngf-mtl/impl.cpp:494]: i prefer to do `*this = other` (and `*this = std::move(other)`  for move ctor)

File: source/ngf-mtl/impl.cpp (hunk 1 above)
Comment (bennywwg) [reply source/ngf-mtl/impl.cpp]: Done

File: source/ngf-mtl/impl.cpp (hunk 2 above)
Comment (bennywwg) [reply source/ngf-mtl/impl.cpp]: Could you explain at a conceptual level why the `std::move` is necessary in the move constructor? Isn't the `other` variable already an 'expiring value' (ie, `&&')?

**Pull request merged**